pkg/resources/management.cattle.io/v3/globalrolebinding/main_test.go

```diff
@@ -0,0 +1,339 @@
+			Name: "admin-cluster-gr",
+		},
+		Rules: []rbacv1.PolicyRule{
+			{
+				APIGroups: []string{""},
+				Resources: []string{"pods"},
+				Verbs:     []string{"get"},
+			},
+		},
+		InheritedClusterRoles: []string{
+			adminRT.Name,
+		},
+	}
+	lockedRoleGR = v3.GlobalRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "locked-gr",
+		},
+		Rules: []rbacv1.PolicyRule{
+			{
+				APIGroups: []string{""},
+				Resources: []string{"pods"},
+				Verbs:     []string{"get"},
+			},
+		},
+		InheritedClusterRoles: []string{
+			lockedRT.Name,
+		},
+	}
+	notFoundRoleGR = v3.GlobalRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "not-found-gr",
+		},
+		Rules: []rbacv1.PolicyRule{
+			{
+				APIGroups: []string{""},
+				Resources: []string{"pods"},
+				Verbs:     []string{"get"},
+			},
+		},
+		InheritedClusterRoles: []string{
+			"not-found",
+		},
+	}
+	errorRoleGR = v3.GlobalRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "error-gr",
+		},
+		Rules: []rbacv1.PolicyRule{
+			{
+				APIGroups: []string{""},
+				Resources: []string{"pods"},
+				Verbs:     []string{"get"},
+			},
+		},
+		InheritedClusterRoles: []string{
+			errName,
+		},
+	}
+)
+
+// createGRBRequest will return a new webhookRequest with the using the given GRBs
+// if oldGRB is nil then a request will be returned as a create operation.
+// if newGRB is nil then a request will be returned as a delete operation.
+// else the request will look like and update operation.
+// if the test.args.username is empty testUser will be used.
+func createGRBRequest(t *testing.T, test testCase) *admission.Request {
+	t.Helper()
+	username := test.args.username
+	if username == "" {
+		username = testUser
+	}
+	gvk := metav1.GroupVersionKind{Group: "management.cattle.io", Version: "v3", Kind: "GlobalRoleBinding"}
+	gvr := metav1.GroupVersionResource{Group: "management.cattle.io", Version: "v3", Resource: "globalrolebindings"}
+	req := &admission.Request{
+		AdmissionRequest: v1.AdmissionRequest{
+			UID:             "1",
+			Kind:            gvk,
+			Resource:        gvr,
+			RequestKind:     &gvk,
+			RequestResource: &gvr,
+			Operation:       v1.Create,
+			UserInfo:        v1authentication.UserInfo{Username: username, UID: ""},
+			Object:          runtime.RawExtension{},
+			OldObject:       runtime.RawExtension{},
+		},
+		Context: context.Background(),
+	}
+	var err error
+	var oldGRB, newGRB *v3.GlobalRoleBinding
+
+	if test.args.newGRB != nil {
+		newGRB = test.args.newGRB()
+	}
+	if test.args.oldGRB != nil {
+		oldGRB = test.args.oldGRB()
+	}
+	if oldGRB != nil {
+		req.Operation = v1.Update
+		req.Name = oldGRB.Name
+		req.Namespace = oldGRB.Namespace
+		req.OldObject.Raw, err = json.Marshal(oldGRB)
+		assert.NoError(t, err, "Failed to marshal GRB while creating request")
+	}
+	if newGRB != nil {
+		req.Name = newGRB.Name
+		req.Namespace = newGRB.Namespace
+		req.Object.Raw, err = json.Marshal(newGRB)
+		assert.NoError(t, err, "Failed to marshal GRB while creating request")
```

Comment:
I don't think this can be combined since one is checking the function, and the other is checking the output of the function. If you can provide an example that doesn't call `test.args.newGRB()` twice I will update.

Reply:
I hadn't considered the case where the function exists but it doesn't return a GRB. I think we can leave it as is then